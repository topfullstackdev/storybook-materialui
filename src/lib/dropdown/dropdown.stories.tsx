import { SelectChangeEvent } from '@mui/material/Select';
import { Meta } from '@storybook/react';
import { useState } from 'react';

import { User, users } from './__tests__/TestDropdownData';
import { Dropdown } from './dropdown';

export default {
  component: Dropdown,
  title: 'Dropdown',
} as Meta;

export function Primary() {
  const [selectedUser, setSelectedUser] = useState(users[0]);

  const extractLabel = (u: User) => u.name;
  const extractKey = (u: User) => u.id;
  const onSelectedUser = (event: SelectChangeEvent, u: User) => {
    setSelectedUser(u);
  };

  return (
    <Dropdown
      value={selectedUser}
      onChange={onSelectedUser}
      label="User Selection"
      options={users}
      extractLabel={extractLabel}
      extractKey={extractKey}
    />
  );
}
