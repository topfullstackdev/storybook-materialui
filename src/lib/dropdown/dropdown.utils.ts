import { SelectChangeEvent } from '@mui/material/Select';

export const onSelectChange = <T>(
  event: SelectChangeEvent<string>,
  options: Readonly<T[]>,
  extractKey: (value: T) => string,
  onChange: (event: SelectChangeEvent<string>, value: T) => void,
) => {
  onChange(
    event,
    options.find(
      (selectedValue) =>
        extractKey(selectedValue).toString() === event.target.value,
    ) as T,
  );
};

export default {
  onSelectChange,
};
