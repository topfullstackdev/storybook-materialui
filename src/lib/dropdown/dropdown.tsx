import { css } from '@emotion/react';
import styled from '@emotion/styled';
import InputBase from '@mui/material/InputBase';
import Select, { SelectChangeEvent, SelectProps } from '@mui/material/Select';

import { useTheme } from '@bai/utils/theme';

import { ReactComponent as DropdownArrow } from '../../../assets/icons/dropdown-arrow.svg';
import { InputLabel } from '../input-label/input-label';

import { onSelectChange } from './dropdown.utils';

export interface DropdownProps<T> extends Omit<SelectProps, 'onChange'> {
  extractLabel: (value: T) => string;
  extractKey: (value: T) => string;
  onChange?: (event: SelectChangeEvent<string>, value: T) => void;
  options: Readonly<T[]>;
  defaultValue?: T;
  value?: T;
  label?: string;
}

const StyledDropdownArrow = function StyledDropdownArrow() {
  return <DropdownArrow height="42px" width="42px" />;
};

export function Dropdown<T>(props: DropdownProps<T>) {
  const {
    options,
    value,
    label,
    onChange,
    extractLabel,
    extractKey,
    defaultValue,
  } = props;

  const theme = useTheme();

  const BootstrapInput = styled(InputBase)(() => ({
    width: '100%',
    'label + &': {
      marginTop: 0,
    },
    '& .MuiInputBase-input': {
      backgroundColor: 'white',
      height: 28,
      width: '100%',
      borderRadius: 0,
      position: 'relative',
      border: `1px solid ${theme.palette.secondary.light}`,
      fontSize: 14,
      fontFamily: 'Verdana, Arial, sans-serif',
      padding: '9px 12px 3px 12px',
      transition:
        'border-color 0.15s ease-in-out 0s, box-shadow 0.15s ease-in-out 0s, -webkit-box-shadow 0.15s ease-in-out 0s',

      '&:focus': {
        borderRadius: 0,
        borderColor: `${theme.palette.primary.main}`,
        boxShadow:
          'rgba(0, 0, 0, 0.075) 0px 1px 1px inset, rgba(102, 175, 233, 0.6) 0px 0px 8px',
      },
    },
    '& .MuiSelect-nativeInput': {
      background: 'red',
    },
  }));

  return (
    <>
      {label && (
        <InputLabel
          css={css`
            margin-bottom: 10px;
          `}
        >
          {label}
        </InputLabel>
      )}
      <Select
        native
        label={label}
        aria-label={label}
        value={(value && extractKey(value)) ?? ''}
        defaultValue={defaultValue && extractKey(defaultValue)}
        onChange={(e) =>
          onChange && onSelectChange(e, options, extractKey, onChange)
        }
        IconComponent={StyledDropdownArrow}
        input={<BootstrapInput role="listbox" aria-label={label} />}
      >
        {options &&
          options.map((option) => (
            <option key={extractKey(option)} value={extractKey(option)}>
              {extractLabel(option)}
            </option>
          ))}
      </Select>
    </>
  );
}

Dropdown.defaultProps = {
  label: null,
};

export default Dropdown;
