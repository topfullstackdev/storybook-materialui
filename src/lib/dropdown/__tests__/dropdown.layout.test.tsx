import { SelectChangeEvent } from '@mui/material/Select';
import { render } from '@testing-library/react';
import { useState } from 'react';

import { Dropdown } from '../dropdown';

import { User, users } from './TestDropdownData';

describe('Dropdown', () => {
  it('should render successfully', () => {
    function DropdownWrapper() {
      const [selectedUser, setSelectedUser] = useState(users[0]);
      const onSelectedUser = (event: SelectChangeEvent, u: User) => {
        setSelectedUser(u);
      };
      return (
        <Dropdown
          value={selectedUser}
          onChange={onSelectedUser}
          label="User Selection"
          options={users}
          extractLabel={(u: User) => u.name}
          extractKey={(u: User) => u.id}
        />
      );
    }
    const { container } = render(<DropdownWrapper />);
    expect(container).toMatchSnapshot();
  });

  it('label should not be rendered', () => {
    function DropdownWrapper() {
      const [selectedUser, setSelectedUser] = useState(users[0]);
      const onSelectedUser = (event: SelectChangeEvent, u: User) => {
        setSelectedUser(u);
      };

      return (
        <Dropdown
          value={selectedUser}
          onChange={onSelectedUser}
          options={users}
          extractLabel={(u: User) => u.name}
          extractKey={(u: User) => u.id}
        />
      );
    }
    const { container } = render(<DropdownWrapper />);
    expect(container).toMatchSnapshot();
  });
});
