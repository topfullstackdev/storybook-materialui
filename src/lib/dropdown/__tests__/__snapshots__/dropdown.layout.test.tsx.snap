// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown label should not be rendered 1`] = `
.emotion-1 {
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.4375em;
  letter-spacing: 0.00938em;
  color: rgba(0, 0, 0, 0.87);
  box-sizing: border-box;
  position: relative;
  cursor: text;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.emotion-1.Mui-disabled {
  color: rgba(0, 0, 0, 0.38);
  cursor: default;
}

label+.emotion-1 {
  margin-top: 0;
}

.emotion-1 .MuiInputBase-input {
  background-color: white;
  height: 28px;
  width: 100%;
  border-radius: 0;
  position: relative;
  border: 1px solid #ba68c8;
  font-size: 14px;
  font-family: Verdana,Arial,sans-serif;
  padding: 9px 12px 3px 12px;
  -webkit-transition: border-color 0.15s ease-in-out 0s,box-shadow 0.15s ease-in-out 0s,-webkit-box-shadow 0.15s ease-in-out 0s;
  transition: border-color 0.15s ease-in-out 0s,box-shadow 0.15s ease-in-out 0s,-webkit-box-shadow 0.15s ease-in-out 0s;
}

.emotion-1 .MuiInputBase-input:focus {
  border-radius: 0;
  border-color: #1976d2;
  box-shadow: rgba(0, 0, 0, 0.075) 0px 1px 1px inset,rgba(102, 175, 233, 0.6) 0px 0px 8px;
}

.emotion-1 .MuiSelect-nativeInput {
  background: red;
}

.emotion-2 {
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 4px;
  cursor: pointer;
  font: inherit;
  letter-spacing: inherit;
  color: currentColor;
  padding: 4px 0 5px;
  border: 0;
  box-sizing: content-box;
  background: none;
  height: 1.4375em;
  margin: 0;
  -webkit-tap-highlight-color: transparent;
  display: block;
  min-width: 0;
  width: 100%;
  -webkit-animation-name: mui-auto-fill-cancel;
  animation-name: mui-auto-fill-cancel;
  -webkit-animation-duration: 10ms;
  animation-duration: 10ms;
}

.emotion-2:focus {
  border-radius: 4px;
}

.emotion-2::-ms-expand {
  display: none;
}

.emotion-2.Mui-disabled {
  cursor: default;
}

.emotion-2[multiple] {
  height: auto;
}

.emotion-2:not([multiple]) option,
.emotion-2:not([multiple]) optgroup {
  background-color: #fff;
}

.emotion-2.emotion-2.emotion-2 {
  padding-right: 32px;
}

.emotion-2::-webkit-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-2::-moz-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-2:-ms-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-2::-ms-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-2:focus {
  outline: 0;
}

.emotion-2:invalid {
  box-shadow: none;
}

.emotion-2::-webkit-search-decoration {
  -webkit-appearance: none;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2::-webkit-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2::-moz-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2:-ms-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2::-ms-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2:focus::-webkit-input-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2:focus::-moz-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2:focus:-ms-input-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2:focus::-ms-input-placeholder {
  opacity: 0.42;
}

.emotion-2.Mui-disabled {
  opacity: 1;
  -webkit-text-fill-color: rgba(0, 0, 0, 0.38);
}

.emotion-2:-webkit-autofill {
  -webkit-animation-duration: 5000s;
  animation-duration: 5000s;
  -webkit-animation-name: mui-auto-fill;
  animation-name: mui-auto-fill;
}

<div>
  <div
    class="MuiInputBase-root MuiInputBase-colorPrimary  emotion-0 emotion-1"
    role="listbox"
    variant="outlined"
  >
    <select
      class="MuiNativeSelect-select MuiNativeSelect-outlined MuiInputBase-input emotion-2"
    >
      <option
        value="1"
      >
        person 1
      </option>
      <option
        value="2"
      >
        person 2
      </option>
      <option
        value="3"
      >
        person 3
      </option>
    </select>
    test-svg-value
  </div>
</div>
`;

exports[`Dropdown should render successfully 1`] = `
.emotion-0 {
  color: rgba(0, 0, 0, 0.6);
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.4375em;
  letter-spacing: 0.00938em;
  padding: 0;
  position: relative;
  display: block;
  transform-origin: top left;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 100%;
  -webkit-transition: color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,-webkit-transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,max-width 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: color 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,transform 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,max-width 200ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  font-size: 1.2rem;
  color: black;
  z-index: 1;
  width: auto;
  margin-bottom: 5px;
  font-family: Verdana,Arial,sans-serif;
  margin-bottom: 10px;
}

.emotion-0.Mui-focused {
  color: #1976d2;
}

.emotion-0.Mui-disabled {
  color: rgba(0, 0, 0, 0.38);
}

.emotion-0.Mui-error {
  color: #d32f2f;
}

.emotion-2 {
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.4375em;
  letter-spacing: 0.00938em;
  color: rgba(0, 0, 0, 0.87);
  box-sizing: border-box;
  position: relative;
  cursor: text;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.emotion-2.Mui-disabled {
  color: rgba(0, 0, 0, 0.38);
  cursor: default;
}

label+.emotion-2 {
  margin-top: 0;
}

.emotion-2 .MuiInputBase-input {
  background-color: white;
  height: 28px;
  width: 100%;
  border-radius: 0;
  position: relative;
  border: 1px solid #ba68c8;
  font-size: 14px;
  font-family: Verdana,Arial,sans-serif;
  padding: 9px 12px 3px 12px;
  -webkit-transition: border-color 0.15s ease-in-out 0s,box-shadow 0.15s ease-in-out 0s,-webkit-box-shadow 0.15s ease-in-out 0s;
  transition: border-color 0.15s ease-in-out 0s,box-shadow 0.15s ease-in-out 0s,-webkit-box-shadow 0.15s ease-in-out 0s;
}

.emotion-2 .MuiInputBase-input:focus {
  border-radius: 0;
  border-color: #1976d2;
  box-shadow: rgba(0, 0, 0, 0.075) 0px 1px 1px inset,rgba(102, 175, 233, 0.6) 0px 0px 8px;
}

.emotion-2 .MuiSelect-nativeInput {
  background: red;
}

.emotion-3 {
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 4px;
  cursor: pointer;
  font: inherit;
  letter-spacing: inherit;
  color: currentColor;
  padding: 4px 0 5px;
  border: 0;
  box-sizing: content-box;
  background: none;
  height: 1.4375em;
  margin: 0;
  -webkit-tap-highlight-color: transparent;
  display: block;
  min-width: 0;
  width: 100%;
  -webkit-animation-name: mui-auto-fill-cancel;
  animation-name: mui-auto-fill-cancel;
  -webkit-animation-duration: 10ms;
  animation-duration: 10ms;
}

.emotion-3:focus {
  border-radius: 4px;
}

.emotion-3::-ms-expand {
  display: none;
}

.emotion-3.Mui-disabled {
  cursor: default;
}

.emotion-3[multiple] {
  height: auto;
}

.emotion-3:not([multiple]) option,
.emotion-3:not([multiple]) optgroup {
  background-color: #fff;
}

.emotion-3.emotion-3.emotion-3 {
  padding-right: 32px;
}

.emotion-3::-webkit-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-3::-moz-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-3:-ms-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-3::-ms-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-3:focus {
  outline: 0;
}

.emotion-3:invalid {
  box-shadow: none;
}

.emotion-3::-webkit-search-decoration {
  -webkit-appearance: none;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3::-webkit-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3::-moz-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3:-ms-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3::-ms-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3:focus::-webkit-input-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3:focus::-moz-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3:focus:-ms-input-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-3:focus::-ms-input-placeholder {
  opacity: 0.42;
}

.emotion-3.Mui-disabled {
  opacity: 1;
  -webkit-text-fill-color: rgba(0, 0, 0, 0.38);
}

.emotion-3:-webkit-autofill {
  -webkit-animation-duration: 5000s;
  animation-duration: 5000s;
  -webkit-animation-name: mui-auto-fill;
  animation-name: mui-auto-fill;
}

<div>
  <label
    class="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary emotion-0"
  >
    User Selection
  </label>
  <div
    aria-label="User Selection"
    class="MuiInputBase-root MuiInputBase-colorPrimary  emotion-1 emotion-2"
    role="listbox"
    variant="outlined"
  >
    <select
      class="MuiNativeSelect-select MuiNativeSelect-outlined MuiInputBase-input emotion-3"
    >
      <option
        value="1"
      >
        person 1
      </option>
      <option
        value="2"
      >
        person 2
      </option>
      <option
        value="3"
      >
        person 3
      </option>
    </select>
    test-svg-value
  </div>
</div>
`;
